In [2]:
import pandas as pd
import os
import glob

# Path to clinical metadata files
clinical_folder = "/Volumes/Jagannath/Projects/Clinical project/new_data/Clinical_meta_data_v1"
clinical_files = glob.glob(os.path.join(clinical_folder, "*_meta.txt"))

clinical_dfs = []
for file in clinical_files:
    # Read each file (assuming tab-delimited)
    df = pd.read_csv(file, sep="\t")
    # Extract cancer type from filename (e.g., BRCA_meta.txt -> BRCA)
    cancer_type = os.path.basename(file).split("_")[0]
    df["Cancer_Type"] = cancer_type
    clinical_dfs.append(df)

# Combine all clinical data into one DataFrame
clinical_data = pd.concat(clinical_dfs, ignore_index=True)

print("Combined Clinical Data:")
print(clinical_data.head())
print("Number of patients in clinical data:", clinical_data.shape[0])

# save the combined clinical data to a new file
output_file = os.path.join(clinical_folder, "combined_clinical_data.txt")
clinical_data.to_csv(output_file, sep="\t", index=False)
print(f"Combined clinical data saved to {output_file}")

Combined Clinical Data:
         idx Tumor Normal  Age     Sex Tumor_Size_cm  \
0  data_type   BIN    BIN  CON     BIN           CON   
1  C3L-00081   Yes    Yes   61  Female             5   
2  C3L-00415   Yes    Yes   64  Female           6.5   
3  C3L-00445   Yes    Yes   74    Male             5   
4  C3L-00568   Yes    Yes   74  Female           4.7   

               Histologic_Grade  Tumor_necrosis Path_Stage_pT Path_Stage_pN  \
0                           ORD             BIN           ORD           ORD   
1  G2 Moderately differentiated  Not identified           pT2           pN1   
2      G3 Poorly differentiated  Not identified           pT2           pN0   
3  G2 Moderately differentiated         Present           pT2           pN0   
4  G2 Moderately differentiated  Not identified           pT2           pN0   

   ... TCF7L2_mutation ZFP36L2_mutation ANO10_mutation MXRA8_mutation  \
0  ...             NaN              NaN            NaN            NaN   
1  ...            

In [4]:
# Path to CNV files
cnv_folder = "/Volumes/Jagannath/Projects/Clinical project/new_data/CNV_WGS_WashU_v1"
cnv_files = glob.glob(os.path.join(cnv_folder, "*.tsv"))

cnv_dfs = []
for file in cnv_files:
    df = pd.read_csv(file, sep="\t")  # adjust sep if needed
    # Extract cancer type from filename (e.g., CPTAC_ccRCC_discovery_CNV_gene_level_log2ratio.tsv -> ccRCC)
    base = os.path.basename(file)
    # This simple extraction assumes the pattern: "CPTAC_<cancer>_discovery_..."
    cancer_type = base.split("_")[1]
    df["Cancer_Type"] = cancer_type
    # Optionally, you might want to set the gene identifier as the index:
    # df.set_index("Gene Symbol", inplace=True)
    cnv_dfs.append(df)

# Combine all CNV data
cnv_data = pd.concat(cnv_dfs, ignore_index=True)

print("Combined CNV Data:")
print(cnv_data.head())
print("CNV data shape:", cnv_data.shape)

# Save the combined CNV data to a new file
output_cnv_file = os.path.join(cnv_folder, "combined_cnv_data.txt")
cnv_data.to_csv(output_cnv_file, sep="\t", index=False)
print(f"Combined CNV data saved to {output_cnv_file}")

Combined CNV Data:
  Unnamed: 0  C3L-00004  C3L-00010  C3L-00011  C3L-00026  C3L-00079  \
0      OR4F5   0.009405   0.010501   0.118637  -0.578567        NaN   
1     OR4F29   0.009405   0.010501   0.118637  -0.578567   0.048056   
2     OR4F16   0.009405   0.010501   0.118637   2.464256   0.048056   
3     SAMD11   0.009405   0.010501   0.118637  -0.067869   0.048056   
4      NOC2L   0.009405   0.010501   0.118637  -0.067869   0.048056   

   C3L-00088  C3L-00096  C3L-00097  C3L-00103  ...  C3N-03851  C3N-03875  \
0   0.141188   0.073666  -0.094625        NaN  ...        NaN        NaN   
1   0.141188   0.073666   1.755733   0.022813  ...        NaN        NaN   
2   2.596901   0.073666  -0.022712   0.022813  ...        NaN        NaN   
3   0.002553   0.073666  -0.022712   0.022813  ...        NaN        NaN   
4   0.002553   0.073666  -0.022712   0.022813  ...        NaN        NaN   

   C3N-03877  C3N-03880  C3N-03882  C3N-03886  C3N-04124  C3N-04127  \
0        NaN        NaN   

In [5]:
# Path to proteome files
proteome_folder = "/Volumes/Jagannath/Projects/Clinical project/new_data/Proteome_BCM_GENCODE_v34_harmonized_v1"
proteome_files = glob.glob(os.path.join(proteome_folder, "*_proteomics_gene_abundance_log2_reference_intensity_normalized_*.txt"))

proteome_dfs = []
for file in proteome_files:
    df = pd.read_csv(file, sep="\t")
    base = os.path.basename(file)
    # Extract cancer type and sample type (Normal/Tumor) from filename
    # Example filename: "BRCA_proteomics_gene_abundance_log2_reference_intensity_normalized_Tumor.txt"
    parts = base.split("_")
    cancer_type = parts[0]
    sample_type = parts[-1].split(".")[0]  # Should be "Tumor" or "Normal"
    df["Cancer_Type"] = cancer_type
    df["Sample_Type"] = sample_type
    proteome_dfs.append(df)

# Combine all proteome data
proteome_data = pd.concat(proteome_dfs, ignore_index=True)

print("Combined Proteome Data:")
print(proteome_data.head())
print("Proteome data shape:", proteome_data.shape)

# Save the combined proteome data to a new file
output_proteome_file = os.path.join(proteome_folder, "combined_proteome_data.txt")
proteome_data.to_csv(output_proteome_file, sep="\t", index=False)
print(f"Combined proteome data saved to {output_proteome_file}")

Combined Proteome Data:
                  idx  C3L-02665  C3L-01663  C3N-02575  C3L-02546  C3L-02963  \
0  ENSG00000000003.15  23.599419  23.756552  23.165015  23.294306  22.615341   
1  ENSG00000000419.12  25.654360  25.500247  25.453226  25.427708  25.242813   
2  ENSG00000000457.14  22.764894  22.752782  22.770509  22.966771  23.350379   
3  ENSG00000000460.17  20.337217  20.277206  19.551387  19.520240        NaN   
4  ENSG00000000938.13  24.872835  23.879392  23.908335  23.582561  23.853372   

   C3N-04162  C3L-02646  C3N-02285  C3N-03875  ...  21BR002  01BR026  05BR004  \
0  22.941939  22.517194  22.887026  23.297662  ...      NaN      NaN      NaN   
1  25.517468  25.228898  25.219671  25.233076  ...      NaN      NaN      NaN   
2  22.820538  22.924718  22.739569  22.575443  ...      NaN      NaN      NaN   
3        NaN        NaN        NaN        NaN  ...      NaN      NaN      NaN   
4  24.395435  23.515016  23.472628  23.562338  ...      NaN      NaN      NaN   

   03BR0

In [6]:
import pandas as pd

# File paths for your combined files
clinical_file = "/Volumes/Jagannath/Projects/Clinical project/new_data/Clinical_meta_data_v1/combined_clinical_data.txt"
cnv_file = "/Volumes/Jagannath/Projects/Clinical project/new_data/CNV_WGS_WashU_v1/combined_cnv_data.txt"
proteome_file = "/Volumes/Jagannath/Projects/Clinical project/new_data/Proteome_BCM_GENCODE_v34_harmonized_v1/combined_proteome_data.txt"

# Load the combined files (assuming they are tab-delimited; adjust sep if needed)
clinical_data = pd.read_csv(clinical_file, sep="\t")
cnv_data = pd.read_csv(cnv_file, sep="\t")
proteome_data = pd.read_csv(proteome_file, sep="\t")

# Inspect the Clinical Data
print("Clinical Data:")
print(clinical_data.head())
print("\nColumns in Clinical Data:")
print(clinical_data.columns)
print("\nNumber of patients in Clinical Data:", clinical_data.shape[0])

# Inspect the CNV Data
print("\nCNV Data:")
print(cnv_data.head())
print("\nColumns in CNV Data:")
print(cnv_data.columns)
print("\nCNV Data shape:", cnv_data.shape)

# Inspect the Proteome Data
print("\nProteome Data:")
print(proteome_data.head())
print("\nColumns in Proteome Data:")
print(proteome_data.columns)
print("\nProteome Data shape:", proteome_data.shape)

# =============================================================================
# Next: Check for common sample identifiers between Clinical and Proteome data.
# =============================================================================

# For the clinical data, let's assume that the column "idx" contains the patient/sample IDs.
clinical_ids = set(clinical_data['idx'].unique())
print("\nSample Clinical IDs (first 10):", list(clinical_ids)[:10])

# For the proteome data, the first column is likely a protein identifier.
# The rest of the columns are the sample IDs. Let's extract them:
proteome_sample_ids = list(proteome_data.columns[1:])  # skip the protein id column
print("\nProteome Sample IDs (first 10):", proteome_sample_ids[:10])

# You might need to adjust the proteome IDs if they include a prefix (e.g., "X").
# Example: remove a leading "X" if present.
def clean_proteome_id(pid):
    return pid[1:] if pid.startswith("X") else pid

cleaned_proteome_ids = set(clean_proteome_id(pid) for pid in proteome_sample_ids)
print("\nCleaned Proteome IDs (first 10):", list(cleaned_proteome_ids)[:10])

# Now, find the intersection between the clinical IDs and the cleaned proteome sample IDs.
common_ids_clinical_proteome = clinical_ids.intersection(cleaned_proteome_ids)
print("\nNumber of patients common to Clinical and Proteome Data:", len(common_ids_clinical_proteome))

# =============================================================================
# CNV Data:
# For CNV data, if it's in wide format, the sample IDs are likely in the columns after the first few.
# Inspect sample columns (assuming the first 3 columns are identifiers like 'Gene Symbol', etc.)
cnv_sample_ids = list(cnv_data.columns[3:])
print("\nCNV Sample IDs (first 10):", cnv_sample_ids[:10])

# At this stage, you need to check whether the CNV sample IDs can be mapped to clinical IDs.
# For example, if they follow a different naming convention, look for a pattern or consult metadata.
# Here we simply print the first few CNV sample IDs for manual inspection.

Clinical Data:
         idx Tumor Normal  Age     Sex Tumor_Size_cm  \
0  data_type   BIN    BIN  CON     BIN           CON   
1  C3L-00081   Yes    Yes   61  Female             5   
2  C3L-00415   Yes    Yes   64  Female           6.5   
3  C3L-00445   Yes    Yes   74    Male             5   
4  C3L-00568   Yes    Yes   74  Female           4.7   

               Histologic_Grade  Tumor_necrosis Path_Stage_pT Path_Stage_pN  \
0                           ORD             BIN           ORD           ORD   
1  G2 Moderately differentiated  Not identified           pT2           pN1   
2      G3 Poorly differentiated  Not identified           pT2           pN0   
3  G2 Moderately differentiated         Present           pT2           pN0   
4  G2 Moderately differentiated  Not identified           pT2           pN0   

   ... TCF7L2_mutation ZFP36L2_mutation ANO10_mutation MXRA8_mutation  \
0  ...             NaN              NaN            NaN            NaN   
1  ...             NaN     

have 1057 common patient/sample IDs between the clinical and proteome datasets. This means that for 1057 patients, the clinical metadata and proteomic measurements are directly linked and can be merged. But not cnv data, cause it has different naming pattern

In [7]:
# 1. Extract patient IDs from each dataset
clinical_ids = set(clinical_data['idx'])
proteome_ids = set(proteome_data.columns[1:])  # Skip the first column (protein IDs)
cnv_ids = set(cnv_data.columns[1:])  # Skip the first column (gene names)

print(f"Number of unique clinical IDs: {len(clinical_ids)}")
print(f"Number of unique proteome patient columns: {len(proteome_ids)}")
print(f"Number of unique CNV patient columns: {len(cnv_ids)}")

Number of unique clinical IDs: 1072
Number of unique proteome patient columns: 1059
Number of unique CNV patient columns: 774


In [8]:
# 2. Standardize IDs across datasets (if needed)
# Look at the format patterns
print("\nID Format Analysis:")
print("Clinical ID examples:", list(clinical_ids)[:5])
print("Proteome ID examples:", list(proteome_ids)[:5])
print("CNV ID examples:", list(cnv_ids)[:5])



ID Format Analysis:
Clinical ID examples: ['C3N-01012', 'C3N-02730', 'C3N-00389', 'C3N-02283', 'C3N-01645']
Proteome ID examples: ['C3N-01012', 'C3N-02730', 'C3N-00389', 'C3L-00921', 'C3N-02283']
CNV ID examples: ['C3N-01012', 'C3N-02730', 'C3N-00389', 'C3L-00921', 'C3N-01645']


In [9]:
# 3. Find overlaps between pairs of datasets
clinical_proteome_overlap = clinical_ids.intersection(proteome_ids)
clinical_cnv_overlap = clinical_ids.intersection(cnv_ids)
proteome_cnv_overlap = proteome_ids.intersection(cnv_ids)

print(f"\nPatients common to Clinical and Proteome: {len(clinical_proteome_overlap)}")
print(f"Patients common to Clinical and CNV: {len(clinical_cnv_overlap)}")
print(f"Patients common to Proteome and CNV: {len(proteome_cnv_overlap)}")


Patients common to Clinical and Proteome: 1057
Patients common to Clinical and CNV: 726
Patients common to Proteome and CNV: 723


In [10]:
# 4. Find IDs common to all three datasets
common_all = clinical_ids.intersection(proteome_ids, cnv_ids)
print(f"\nPatients common to all three datasets: {len(common_all)}")

# 5. If we need to normalize ID formats, create mapping functions
def normalize_id(id_string):
    """Normalize ID formats across datasets"""
    # Remove any spaces or special characters
    id_clean = id_string.strip()
    
    # Handle any specific transformations needed
    # (This is just an example - adjust based on your actual ID patterns)
    if '-' not in id_clean and len(id_clean) > 5:
        # Insert hyphen if missing (e.g., C3L00011 -> C3L-00011)
        id_clean = id_clean[:3] + '-' + id_clean[3:]
    
    return id_clean

# Apply normalization if needed (uncomment and adapt as necessary)
# normalized_clinical_ids = {normalize_id(id) for id in clinical_ids}
# normalized_proteome_ids = {normalize_id(id) for id in proteome_ids}
# normalized_cnv_ids = {normalize_id(id) for id in cnv_ids}


Patients common to all three datasets: 722


The patient IDs are already consistent across all three datasets, so we don't need to perform any ID reformatting. With 722 patients having data in all three datasets, we have an excellent foundation for our analysis.

In [12]:
# 1. Start with the common patient list
common_patients = list(clinical_ids.intersection(proteome_ids, cnv_ids))

# 2. Create core clinical dataset
core_clinical = clinical_data[clinical_data['idx'].isin(common_patients)].copy()

# 3. Select relevant clinical features (adjust based on what's available in your data)
key_clinical_features = [
    'idx', 'Age', 'Sex', 'Tumor_Size_cm', 'Histologic_Grade', 
    'Path_Stage_pT', 'Path_Stage_pN'
]

In [13]:
# Add survival and outcome features if available
if 'OS_days' in clinical_data.columns and 'OS_event' in clinical_data.columns:
    key_clinical_features.extend(['OS_days', 'OS_event'])
# Add any mutation features that are well-populated
mutation_cols = [col for col in clinical_data.columns if '_mutation' in col]
for col in mutation_cols:
    # Only include mutations with adequate data
    if clinical_data[col].notna().mean() > 0.5:  # More than 50% complete
        key_clinical_features.append(col)

In [14]:
# Filter to existing columns
existing_features = [f for f in key_clinical_features if f in core_clinical.columns]
core_clinical = core_clinical[existing_features]

# 4. Save the core clinical dataset
core_clinical.to_csv('core_clinical_data.csv', index=False)
print(f"Core clinical dataset saved with {core_clinical.shape[0]} patients and {core_clinical.shape[1]} features")

Core clinical dataset saved with 722 patients and 11 features


## Database

In [19]:
import sqlite3
import pandas as pd
import numpy as np
import os

# First, try to remove any existing database to start fresh
db_path = 'clinical_trial_monitor.db'
if os.path.exists(db_path):
    try:
        os.remove(db_path)
        print(f"Removed existing database: {db_path}")
    except:
        print(f"Could not remove database: {db_path}. It may be in use by another process.")

# Create a new database connection
conn = sqlite3.connect(db_path)
conn.isolation_level = None  # This enables autocommit mode
cursor = conn.cursor()

# Begin transaction
cursor.execute('BEGIN TRANSACTION')

# Create the patients table
cursor.execute('''
CREATE TABLE IF NOT EXISTS patients (
    patient_id TEXT PRIMARY KEY,
    age INTEGER,
    sex TEXT,
    tumor_size_cm REAL,
    histologic_grade TEXT,
    path_stage_pt TEXT,
    path_stage_pn TEXT,
    os_days REAL,
    os_event INTEGER,
    pik3ca_mutation INTEGER,
    tp53_mutation INTEGER
)
''')

# Create other tables as before
cursor.execute('''
CREATE TABLE IF NOT EXISTS proteomics (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    patient_id TEXT,
    protein_id TEXT,
    expression_level REAL,
    FOREIGN KEY (patient_id) REFERENCES patients(patient_id)
)
''')

cursor.execute('''
CREATE TABLE IF NOT EXISTS genomics (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    patient_id TEXT,
    gene_id TEXT,
    copy_number_variation REAL,
    FOREIGN KEY (patient_id) REFERENCES patients(patient_id)
)
''')

cursor.execute('''
CREATE TABLE IF NOT EXISTS adverse_events (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    patient_id TEXT,
    event_date TEXT,
    event_type TEXT,
    event_grade INTEGER,
    related_to_treatment INTEGER,
    FOREIGN KEY (patient_id) REFERENCES patients(patient_id)
)
''')

# Commit the table creation
cursor.execute('COMMIT')
print("Tables created successfully")

# Load the clinical data
core_clinical = pd.read_csv('core_clinical_data.csv')

# Prepare the data as a list of tuples for more efficient loading
records = []
for _, row in core_clinical.iterrows():
    record = (
        row['idx'],
        int(row['Age']) if pd.notna(row['Age']) else None,
        row['Sex'],
        float(row['Tumor_Size_cm']) if pd.notna(row['Tumor_Size_cm']) else None,
        row['Histologic_Grade'],
        row['Path_Stage_pT'],
        row['Path_Stage_pN'],
        float(row['OS_days']) if pd.notna(row['OS_days']) else None,
        int(row['OS_event']) if pd.notna(row['OS_event']) else None,
        int(row['PIK3CA_mutation']) if pd.notna(row['PIK3CA_mutation']) else None,
        int(row['TP53_mutation']) if pd.notna(row['TP53_mutation']) else None
    )
    records.append(record)

# Insert in batches using executemany
try:
    cursor.execute('BEGIN TRANSACTION')
    cursor.executemany('''
    INSERT INTO patients VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    ''', records)
    cursor.execute('COMMIT')
    print(f"Successfully inserted {len(records)} patients into the database")
except sqlite3.Error as e:
    print(f"Database error: {e}")
    cursor.execute('ROLLBACK')

# Close the connection properly
cursor.close()
conn.close()
print("Database connection closed")

Removed existing database: clinical_trial_monitor.db
Tables created successfully
Successfully inserted 722 patients into the database
Database connection closed


In [20]:
import sqlite3
import pandas as pd
import numpy as np

# Connect to the database
conn = sqlite3.connect('clinical_trial_monitor.db')
cursor = conn.cursor()

# 1. Load and transform proteomics data
# Get the list of patients we have in the database
cursor.execute("SELECT patient_id FROM patients")
patient_ids = [row[0] for row in cursor.fetchall()]

# Load the proteome data
proteome_data = pd.read_csv("/Volumes/Jagannath/Projects/Clinical project/new_data/Proteome_BCM_GENCODE_v34_harmonized_v1/combined_proteome_data.txt", sep="\t")

# Find the intersection of patient IDs in our DB and in the proteome columns
proteome_columns = [col for col in proteome_data.columns if col in patient_ids]
proteome_subset = proteome_data[['idx'] + proteome_columns]

# Select top 100 proteins based on variance (for demonstration)
protein_variance = proteome_subset[proteome_columns].var(axis=1)
top_proteins = protein_variance.nlargest(100).index

# Transform from wide to long format for the selected proteins
records = []
for idx in top_proteins:
    protein_id = proteome_subset.iloc[idx, 0]  # Get protein ID from idx column
    for patient_id in proteome_columns:
        expression = proteome_subset.iloc[idx, proteome_subset.columns.get_loc(patient_id)]
        if pd.notna(expression):
            records.append((patient_id, protein_id, float(expression)))

# Insert in batches
print(f"Inserting {len(records)} proteomics records...")
cursor.execute('BEGIN TRANSACTION')
cursor.executemany('''
INSERT INTO proteomics (patient_id, protein_id, expression_level)
VALUES (?, ?, ?)
''', records)
conn.commit()
print("Proteomics data loaded successfully")

# 2. Load and transform CNV data
# Load the CNV data
cnv_data = pd.read_csv("/Volumes/Jagannath/Projects/Clinical project/new_data/CNV_WGS_WashU_v1/combined_cnv_data.txt", sep="\t")

# Find the intersection of patient IDs in our DB and in the CNV columns
cnv_columns = [col for col in cnv_data.columns if col in patient_ids]
cnv_subset = cnv_data[['Unnamed: 0'] + cnv_columns]  # Assuming first column has gene IDs

# Select top 100 genes based on variance
gene_variance = cnv_subset[cnv_columns].var(axis=1)
top_genes = gene_variance.nlargest(100).index

# Transform from wide to long format for the selected genes
records = []
for idx in top_genes:
    gene_id = cnv_subset.iloc[idx, 0]  # Get gene ID from first column
    for patient_id in cnv_columns:
        cnv_value = cnv_subset.iloc[idx, cnv_subset.columns.get_loc(patient_id)]
        if pd.notna(cnv_value):
            records.append((patient_id, gene_id, float(cnv_value)))

# Insert in batches
print(f"Inserting {len(records)} genomics records...")
cursor.execute('BEGIN TRANSACTION')
cursor.executemany('''
INSERT INTO genomics (patient_id, gene_id, copy_number_variation)
VALUES (?, ?, ?)
''', records)
conn.commit()
print("Genomics data loaded successfully")

# 3. Generate simulated adverse events for demonstration
import random
from datetime import datetime, timedelta

# Types of adverse events in clinical trials
event_types = [
    "Fever", "Nausea", "Fatigue", "Headache", "Skin rash", 
    "Neutropenia", "Anemia", "Thrombocytopenia", "Vomiting", "Diarrhea"
]

# Generate random dates within the past year
start_date = datetime.now() - timedelta(days=365)
end_date = datetime.now()

# Generate records
records = []
for patient_id in patient_ids[:200]:  # Generate events for a subset of patients
    # Each patient may have 0-3 adverse events
    num_events = random.randint(0, 3)
    for _ in range(num_events):
        event_type = random.choice(event_types)
        event_grade = random.randint(1, 3)  # Grade 1-3
        related = random.choice([0, 1])  # 0=Not related, 1=Related
        
        # Random date
        days_between = (end_date - start_date).days
        random_day = random.randint(0, days_between)
        event_date = (start_date + timedelta(days=random_day)).strftime('%Y-%m-%d')
        
        records.append((patient_id, event_date, event_type, event_grade, related))

# Insert the records
print(f"Inserting {len(records)} adverse event records...")
cursor.execute('BEGIN TRANSACTION')
cursor.executemany('''
INSERT INTO adverse_events (patient_id, event_date, event_type, event_grade, related_to_treatment)
VALUES (?, ?, ?, ?, ?)
''', records)
conn.commit()
print("Adverse events data loaded successfully")

# Close the connection
cursor.close()
conn.close()
print("Database connection closed")

Inserting 2911 proteomics records...
Proteomics data loaded successfully
Inserting 9500 genomics records...
Genomics data loaded successfully
Inserting 282 adverse event records...
Adverse events data loaded successfully
Database connection closed


Dashboard Development
Basic SQL Queries for Dashboard Components

In [21]:
import sqlite3
import pandas as pd

# Connect to database
conn = sqlite3.connect('clinical_trial_monitor.db')

# 1. Patient demographics overview
demographics = pd.read_sql('''
SELECT 
    sex, 
    COUNT(*) as count,
    AVG(age) as avg_age,
    MIN(age) as min_age,
    MAX(age) as max_age
FROM patients
GROUP BY sex
''', conn)
print("Patient Demographics:")
print(demographics)

# 2. Tumor characteristics 
tumor_stats = pd.read_sql('''
SELECT 
    histologic_grade,
    COUNT(*) as count,
    AVG(tumor_size_cm) as avg_size
FROM patients
WHERE histologic_grade IS NOT NULL
GROUP BY histologic_grade
''', conn)
print("\nTumor Characteristics:")
print(tumor_stats)

# 3. Survival by mutation status
survival = pd.read_sql('''
SELECT 
    tp53_mutation,
    COUNT(*) as patient_count,
    SUM(os_event) as events,
    AVG(os_days) as avg_survival_days
FROM patients
WHERE tp53_mutation IS NOT NULL
GROUP BY tp53_mutation
''', conn)
print("\nSurvival by TP53 Mutation Status:")
print(survival)

# 4. Most common adverse events
adverse_events = pd.read_sql('''
SELECT 
    event_type,
    COUNT(*) as event_count,
    AVG(event_grade) as avg_severity
FROM adverse_events
GROUP BY event_type
ORDER BY event_count DESC
LIMIT 5
''', conn)
print("\nTop 5 Adverse Events:")
print(adverse_events)

conn.close()

Patient Demographics:
      sex  count    avg_age  min_age  max_age
0  Female    283  62.491166       24       90
1    Male    439  62.380410       23       89

Tumor Characteristics:
               histologic_grade  count  avg_size
0        G1 Well differentiated     71  3.729577
1  G2 Moderately differentiated    353  3.890571
2      G3 Poorly differentiated    184  4.833702
3           G4 Undifferentiated     10  9.110000

Survival by TP53 Mutation Status:
   tp53_mutation  patient_count  events  avg_survival_days
0              0            222      79         818.507177
1              1            395     156         782.698630

Top 5 Adverse Events:
         event_type  event_count  avg_severity
0  Thrombocytopenia           34      2.176471
1             Fever           33      1.848485
2          Vomiting           31      1.806452
3          Diarrhea           30      2.233333
4         Skin rash           29      2.137931
